ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/DistributingExecutor.java

```diff
@@ -64,18 +67,26 @@ public DistributingExecutor(
       final ConfiguredStatement<Statement> statement,
       final Map<String, Object> mutableScopedProperties,
       final KsqlExecutionContext executionContext,
-      final ServiceContext serviceContext) {
+      final ServiceContext serviceContext
+  ) {
     final ConfiguredStatement<?> injected = injectorFactory
         .apply(executionContext, serviceContext)
         .inject(statement);
 
     checkAuthorization(injected, serviceContext, executionContext);
 
     try {
-      final QueuedCommandStatus queuedCommandStatus = commandQueue.enqueueCommand(injected);
+      if (transactionalProducer == null) {
+        throw new RuntimeException("Transaction manager for distributing executor not set");
+      }
+
+      final QueuedCommandStatus queuedCommandStatus =
+          commandQueue.enqueueCommand(injected, transactionalProducer);
+
       final CommandStatus commandStatus = queuedCommandStatus
```

Comment:
We could also re-run validation for logged statements as part of execute. So `DistributingExecutor` would just run the whole protocol (init txn, validate, log cmd, commit). It's a bit wasteful to re-run validation, but I think this is the simplest way to get this PR unblocked.

Reply:
> - run the validate-execute-commit loop for each statement one-by-one (rather than validating them all up front).

I think we have to go with this option. The first option could potentially return misleading results in the event of an abortTransaction();

```
//starting from no streams
CREATE STREAM foo1(age BIGINT) WITH (KAFKA_TOPIC='foo',  VALUE_FORMAT='JSON');
CREATE STREAM foo2(age BIGINT) WITH (KAFKA_TOPIC='foo',  VALUE_FORMAT='JSON');
show streams;
```
If validations passed but when trying to commit the 2 records, it fails and the transaction is aborted, those streams wouldn't be created, but the show streams in that batch of statements would show that Stream foo1 and foo2 were created since it would be successfully executed against a sandboxEngine and then the `show streams` is run against a sandboxEngine with those two streams. However, running show streams again would show no streams have been created.

Reply:
> - run the validate-execute-commit loop for each statement one-by-one (rather than validating them all up front).

I think there's also issues with this. With how the code is currently designed, a KsqlEntityList is only returned if all the statements are successfully executed.

```
ksql> CREATE TABLE foo400(age BIGINT) WITH (KAFKA_TOPIC='foo',  VALUE_FORMAT='JSON');
>CREATE TABLE foo200(age BIGINT) WITH (KAFKA_TOPIC='foo',  VALUE_FORMAT='JSON');
>Show Tables;

CREATE TABLE foo400(age BIGINT) WITH (KAFKA_TOPIC='foo',  VALUE_FORMAT='JSON');
 Message       
---------------
 Table created 
---------------

CREATE TABLE foo200(age BIGINT) WITH (KAFKA_TOPIC='foo',  VALUE_FORMAT='JSON');
 Message       
---------------
 Table created 
---------------

Show Tables;
 Table Name | Kafka Topic | Format | Windowed 
----------------------------------------------
 FOO200     | foo         | JSON   | false    
 FOO400     | foo         | JSON   | false    
----------------------------------------------
ksql> 
ksql> CREATE TABLE foo100(age BIGINT) WITH (KAFKA_TOPIC='foo',  VALUE_FORMAT='JSON');
>CREATE TABLE foo200(age BIGINT) WITH (KAFKA_TOPIC='foo',  VALUE_FORMAT='JSON');
>Show Tables;
Cannot add table 'FOO200': A table with the same name already exists
ksql> show tables;

 Table Name | Kafka Topic | Format | Windowed 
----------------------------------------------
 FOO100     | foo         | JSON   | false    
 FOO200     | foo         | JSON   | false    
 FOO400     | foo         | JSON   | false    
----------------------------------------------
ksql> 
```

The code would need to be refactored to be able to handle processing individual commands (which should probably be done before implementing the TransactionalProduces)

Reply:
> This option could potentially return misleading results in the event of an abortTransaction();
In this case we'd just fail the whole request - so no response to `list streams;`

Reply:
> The code would need to be refactored to be able to handle processing individual commands (which should probably be done before implementing the TransactionalProduces)

I'm not sure I follow. It's possible today for any command to fail when executing, in which case we just fail the whole command. We could do the same thing if validation inside the transaction fails. I agree it's better to return the partial response with the results for what we could successfully execute. My point is that this is orthogonal to what we're trying to do here.

Reply:
> I agree it's better to return the partial response with the results for what we could successfully execute. My point is that this is orthogonal to what we're trying to do here.

I also think it would be good to have a partial response, but it's not a deal breaker for me. If the team is OK with the behavior I outlined here https://github.com/confluentinc/ksql/pull/3660#discussion_r341442966 then I'm on board.

Reply:
I ended up going with  

Running the protocol on each statement one-by-one (re running validation on distributed statements)

Reply:
Yeah doing the validation in `DistributedExecutor` was just meant to be an immediate-term thing so we can get this change merged without blocking on deciding how we really want the validation to be done. I'm personally OK with validating-executing each statement one by one. This is the way it was done at first. At some point @big-andy-coates changed it to validate everything up-front. It would be good to get his perspective on why this was done.